File: reset-samples/go/reset-point/workflow.go
package main

import (
	"context"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/workflow"
)

// OrderWorkflow demonstrates using reset points for workflow checkpointing
func OrderWorkflow(ctx workflow.Context, orderID string) (string, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("Order workflow started", "orderID", orderID)

	// Set activity options
	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 10 * time.Second,
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	// Step 1: Validate order
	var validationResult string
	err := workflow.ExecuteActivity(ctx, ValidateOrder, orderID).Get(ctx, &validationResult)
	if err != nil {
		return "", err
	}
	logger.Info("Order validated", "result", validationResult)

	// RESET POINT: Can reset to here if payment fails
	err = ResetPoint(ctx, "after-validation")
	if err != nil {
		logger.Error("Failed to record reset point", "error", err)
		return "", err
	}
	logger.Info("Recorded reset point: after-validation")

	// Step 2: Process payment
	var paymentResult string
	err = workflow.ExecuteActivity(ctx, ProcessPayment, orderID).Get(ctx, &paymentResult)
	if err != nil {
		return "", err
	}
	logger.Info("Payment processed", "result", paymentResult)

	// RESET POINT: Can reset to here if shipping fails
	err = ResetPoint(ctx, "after-payment")
	if err != nil {
		logger.Error("Failed to record reset point", "error", err)
		return "", err
	}
	logger.Info("Recorded reset point: after-payment")

	// Step 3: Ship order
	var shippingResult string
	err = workflow.ExecuteActivity(ctx, ShipOrder, orderID).Get(ctx, &shippingResult)
	if err != nil {
		return "", err
	}
	logger.Info("Order shipped", "result", shippingResult)

	return "Order completed: " + orderID, nil
}

// ValidateOrder activity
func ValidateOrder(ctx context.Context, orderID string) (string, error) {
	activity.GetLogger(ctx).Info("Validating order", "orderID", orderID)
	// Simulate validation work
	time.Sleep(100 * time.Millisecond)
	return "Validated-" + orderID, nil
}

// ProcessPayment activity
func ProcessPayment(ctx context.Context, orderID string) (string, error) {
	activity.GetLogger(ctx).Info("Processing payment", "orderID", orderID)
	// Simulate payment processing
	time.Sleep(100 * time.Millisecond)
	return "Paid-" + orderID, nil
}

// ShipOrder activity
func ShipOrder(ctx context.Context, orderID string) (string, error) {
	activity.GetLogger(ctx).Info("Shipping order", "orderID", orderID)
	// Simulate shipping
	time.Sleep(100 * time.Millisecond)
	return "Shipped-" + orderID, nil
}

File: reset-samples/python/reset-point/workflow.py
"""Order workflow demonstrating reset points."""
import asyncio
from datetime import timedelta
from temporalio import workflow

# Import our activities and reset point helper
with workflow.unsafe.imports_passed_through():
    from activities import validate_order, process_payment, ship_order
    from reset_point_helper import record_reset_point


@workflow.defn
class OrderWorkflow:
    """
    Demonstrates using reset points for workflow checkpointing.
    
    Reset points allow you to reset a workflow to a specific named point
    in its execution history.
    """

    @workflow.run
    async def run(self, order_id: str) -> str:
        workflow.logger.info(f"Order workflow started for {order_id}")

        # Step 1: Validate order
        validation_result = await workflow.execute_activity(
            validate_order,
            order_id,
            start_to_close_timeout=timedelta(seconds=10),
        )
        workflow.logger.info(f"Order validated: {validation_result}")

        # RESET POINT
        await record_reset_point("after-validation")

        # Step 2: Process payment
        payment_result = await workflow.execute_activity(
            process_payment,
            order_id,
            start_to_close_timeout=timedelta(seconds=10),
        )
        workflow.logger.info(f"Payment processed: {payment_result}")

        # RESET POINT
        await record_reset_point("after-payment")

        # Step 3: Ship order
        shipping_result = await workflow.execute_activity(
            ship_order,
            order_id,
            start_to_close_timeout=timedelta(seconds=10),
        )
        workflow.logger.info(f"Order shipped: {shipping_result}")

        return f"Order completed: {order_id}"

